### ISSUE TITLE
Make provider reference

### ISSUE THREAD

**Issue by rkdud007** [glihm/starknet-abigen-rs#14]
### Problem 
Right now when we call `new_caller` or `new_invoker`, we pass provider ownership. Which cause us to recreate multiple providers when we initiate contract. 
```rust

    let provider =
        AnyProvider::JsonRpcHttp(JsonRpcClient::new(HttpTransport::new(rpc_url.clone())));

    ...
    // Call.
    let contract_caller = MyContract::new_caller(contract_address, provider).await?;
    let val = contract_caller.get_val().await?;

    // Work in progress to avoid this duplication.
    let provider2 =
        AnyProvider::JsonRpcHttp(JsonRpcClient::new(HttpTransport::new(rpc_url.clone())));

    ...
    // Invoke.
    let contract_invoker =
        MyContract::new_invoker(contract_address, provider2, account_address, signer).await?;
    contract_invoker.set_val(FieldElement::TWO).await?;
```

### Possible approach
Pass provider into reference so that we can use one provider for caller and invoker. 

We might can create init struct like this and pass it.
```rust
ContractSetup {
   address,
   provider,
   Option<account_address>,
   Option<private_key>
}
```

**Comment by glihm** [comment]
Or as discusses @rkdud007 if we only have one `new` function that takes all arguments, we can still taking ownership of the provider. Or do you think it still better to switch to reference?

**Comment by rkdud007** [comment]
What we will do is we bind provider lifetime to Contract lifetime. Contract will have new functiom to initialize -- but account will be optional, and we add with_account function to add account param later :)

**Comment by glihm** [comment]
I do like the idea of `with_account` that can be called later!

**Comment by glihm** [comment]
closed by #15
